tests/basic/mutex.rs

```diff
@@ -114,6 +115,230 @@ fn concurrent_increment() {
+                let lock = Arc::clone(&lock);
+                let value = Arc::clone(&value);
+                thread::spawn(move || {
+                    {
+                        let _guard = lock.lock().unwrap();
+                        *value.lock().unwrap() += 1;
+                    }
+                    {
+                        let _guard = rwlock.write().unwrap();
+                        *value.lock().unwrap() += 2;
+                    }
+                    {
+                        let _guard = lock.lock().unwrap();
+                        *value.lock().unwrap() += 3;
+                    }
+                })
+            };
+            let mul_thread = {
+                let lock = Arc::clone(&lock);
+                let log = Arc::clone(&value);
+                thread::spawn(move || {
+                    let _guard = lock.lock().unwrap();
+                    *log.lock().unwrap() *= 2;
+                })
+            };
+
+            add_thread.join().unwrap();
+            mul_thread.join().unwrap();
+
+            let value = Arc::try_unwrap(value).unwrap().into_inner().unwrap();
+            observed_values_clone.lock().unwrap().insert(value);
+        },
+        None,
+    );
+
+    let observed_values = Arc::try_unwrap(observed_values).unwrap().into_inner().unwrap();
+    assert_eq!(observed_values.len(), 4);
+    assert!(observed_values.contains(&6));
+    assert!(observed_values.contains(&7));
+    assert!(observed_values.contains(&9));
+    assert!(observed_values.contains(&12));
+}
+
+/// Two concurrent threads trying to do an atomic increment using `try_lock`.
+/// One `try_lock` must succeed, while the other may or may not succeed.
+/// Thus we expect to see final values 1 and 2.
+#[test]
+fn concurrent_try_increment() {
+    let observed_values = Arc::new(std::sync::Mutex::new(HashSet::new()));
+    let observed_values_clone = Arc::clone(&observed_values);
+
+    check_dfs(
+        move || {
+            let lock = Arc::new(Mutex::new(0usize));
+
+            let threads = (0..2)
+                .map(|_| {
+                    let lock = Arc::clone(&lock);
+                    thread::spawn(move || {
+                        match lock.try_lock() {
+                            Ok(mut guard) => {
+                                *guard += 1;
+                            }
+                            Err(TryLockError::WouldBlock) => (),
+                            Err(_) => panic!("unexpected TryLockError"),
+                        };
+                    })
+                })
+                .collect::<Vec<_>>();
+
+            for thd in threads {
+                thd.join().unwrap();
+            }
+
+            let value = *lock.try_lock().unwrap();
+            observed_values_clone.lock().unwrap().insert(value);
+        },
+        None,
+    );
+
+    let observed_values = Arc::try_unwrap(observed_values).unwrap().into_inner().unwrap();
+    assert_eq!(observed_values.len(), 2);
```

Comment:
Think you can just do `assert_eq!(observed_values, HashSet::from([1, 2]))` here, and similarly in other tests below.

Reply:
Right, that'll give a more comprehensive error message.